trin-core/src/types/accumulator.rs

```diff
@@ -0,0 +1,203 @@
+            },
+            current_epoch: EpochAccumulator {
+                header_records: HeaderRecordList::empty(),
+            },
+        }
+    }
+
+    // as defined in:
+    // https://github.com/ethereum/portal-network-specs/blob/e807eb09d2859016e25b976f082735d3aceceb8e/history-network.md#the-header-accumulator
+    fn update_accumulator(&mut self, new_block_header: &Header) {
+        // get the previous total difficulty
+        let last_total_difficulty = match self.current_epoch.header_records.len() {
+            // genesis
+            0 => U256::zero(),
+            _ => {
+                self.current_epoch
+                    .header_records
+                    .last()
+                    .unwrap()
+                    .total_difficulty
+            }
+        };
+
+        // check if the epoch accumulator is full
+        if self.current_epoch.header_records.len() == EPOCH_SIZE {
+            // compute the final hash for this epoch
+            let epoch_hash = self.current_epoch.tree_hash_root();
+            // append the hash for this epoch to the list of historical epochs
+            self.historical_epochs
+                .epochs
+                .push(epoch_hash)
+                .expect("Invalid accumulator state, more historical epochs than allowed.");
+            // initialize a new empty epoch
+            self.current_epoch = EpochAccumulator {
+                header_records: HeaderRecordList::empty(),
+            };
+        }
+
+        // construct the concise record for the new header and add it to the current epoch
+        let header_record = HeaderRecord {
+            block_hash: new_block_header.hash(),
+            total_difficulty: last_total_difficulty + new_block_header.difficulty,
+        };
+        self.current_epoch
+            .header_records
+            .push(header_record)
+            .expect("Invalid accumulator state, more current epochs than allowed.");
+    }
+}
+
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct HistoricalEpochs {
+    epochs: HistoricalEpochsList,
+}
+
+impl TreeHash for HistoricalEpochs {
+    fn tree_hash_type() -> tree_hash::TreeHashType {
+        tree_hash::TreeHashType::List
+    }
+
+    fn tree_hash_packed_encoding(&self) -> Vec<u8> {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_packing_factor() -> usize {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_root(&self) -> tree_hash::Hash256 {
+        // list of composite objects
+        // mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value))
+        let hash_tree_roots: Vec<tree_hash::Hash256> = self
+            .clone()
+            .epochs
+            .into_iter()
+            .map(|epoch_root| epoch_root.tree_hash_root())
```

Comment:
Something feels a little funny here about cloning self instead of the epochs. Actually, can the clone be skipped entirely like this?
```suggestion
        let hash_tree_roots: Vec<tree_hash::Hash256> = self
            .epochs
            .iter()
            .map(|epoch_root| epoch_root.tree_hash_root())
```

Ditto on `EpochAccumulator`